src/components/plots/helpers/heatmap/populateHeatmapData.js

```diff
@@ -1,10 +1,12 @@
 import _ from 'lodash';
-import SetOperations from '../../../utils/setOperations';
-import { union } from '../../../utils/cellSetOperations';
+import generateVitessceHeatmapTracksData from 'components/plots/helpers/heatmap/generateVitessceHeatmapTracksData';
+
+import SetOperations from 'utils/setOperations';
+import { union } from 'utils/cellSetOperations';
 
 const populateHeatmapData = (
```

Comment:
have you thought about refactoring this function, splitting it into 2: one for generating vitessce heatmap data and one for non-vitessce heatmap data?

Reply:
It is a 270 lines of function that includes nested functions, it will benefit a lot from some better re-organisation

Reply:
refactoring is another matter, but I don't know if we should do everything in this PR, would refactoring the non-vitessce specific parts of `populateHeatmapData` be part this ticket?

Reply:
could be a good opportunity to do it, but I understand that this PR has been on for a long while and we need to finish it off.

Reply:
Leaving it up to you, if you think it is going to take more effort, then leave it for other time

Reply:
Much better than before! A small thing you can do to make the code more readable: move line 33 `const trackOrder = Array.from(selectedTracks).reverse();` to be just before lines 149 and 150:

```  const cellOrder = generateCellOrder(groupedTracks);
  const geneOrder = selectedGenes;
```

In that way, you have the `populateHeatmapData` function clearly separated into two sections: section with all the inner functions used within it, and section that triggers the actual executions of these functions.